### PR TITLE
feat(graphql): provide secret token used in policy

### PR DESCRIPTION
**Pull request by moreal**
In this pull request, it started to provide a new option `--graphql-secret-token-path`. Through this option, you can use `LocalPolicy` protection (i.e., it forces you to request with `Authorization` header when using `mutation` operation.), and you can set the path of the secret token stored manually.

For example, you can pass it to the header with GraphQL Playground.

image

Also, it will be useful when applying this feature to the frontend using *Apollo*

https://www.apollographql.com/docs/react/networking/advanced-http-networking/

### CODE REVIEW

File: NineChronicles.Headless.Executable/Program.cs [hunk 1]
```diff
@@ -190,12 +195,22 @@ static void ConfigureSentryOptions(SentryOptions o)
...
+                        await File.WriteAllTextAsync(graphQLSecretTokenPath, secretToken);
+                    }
                     var graphQLNodeServiceProperties = new GraphQLNodeServiceProperties
                     {
                         GraphQLServer = graphQLServer,
                         GraphQLListenHost = graphQLHost,
                         GraphQLListenPort = graphQLPort,
+                        SecretToken = secretToken,
                     };
+                    Console.Error.WriteLine(secretToken);
```
Comment (longfin) [inline NineChronicles.Headless.Executable/Program.cs:213]: Is this output necessary?

File: NineChronicles.Headless/GraphQLService.cs [hunk 2]
```diff
@@ -59,15 +81,58 @@ public void ConfigureServices(IServiceCollection services)
...
                     .AddGraphTypes(typeof(StandaloneSchema));
+                services.AddHttpContextAccessor()
+                    .AddTransient<IValidationRule, AuthorizationValidationRule>()
+                    .AddSingleton<IAuthorizationEvaluator, AuthorizationEvaluator>()
+                    .AddTransient(
+                        s =>
+                        {
+                            var authSettings = new AuthorizationSettings();
+                            authSettings.AddPolicy(
+                                "LocalPolicy",
```
Comment (longfin) [inline NineChronicles.Headless/GraphQLService.cs:127]: Can we define constant for "LocalPolicy" ?

File: NineChronicles.Headless.Executable/Program.cs (hunk 1 above)
Comment (moreal) [reply NineChronicles.Headless.Executable/Program.cs]: It was my mistake that remained after debugging and I amended it. Thanks!

File: NineChronicles.Headless/GraphQLService.cs (hunk 2 above)
Comment (moreal) [reply NineChronicles.Headless/GraphQLService.cs]: I moved constant strings such you commented, to `const` variables in `GraphQLService` in c06f870. 🙏

**Review by longfin — APPROVED** [review]

**Review by limebell — APPROVED** [review]

**Review by riemannulus — APPROVED** [review]

**Comment by moreal** [comment]
I changed that the secret token to need also at some HTTP endpoints and some GraphQL queries. cf7d441, d8a13c1. Please review them again @riemannulus @longfin @limebell 🙇 

## GraphQL queries
- keyStore
- activationStatus

## HTTP endpoints
- POST /run-standalone
- POST /set-mining
- POST /set-private-key

File: NineChronicles.Headless/Controllers/GraphQLController.cs [hunk 3]
```diff
@@ -261,5 +282,16 @@ private void NotifyAction(ActionBase.ActionEvaluation<ActionBase> eval)
                 }
             }
         }
+
+        // FIXME: remove this method with DI.
+        private IActionResult AuthenticateLocalPolicy()
```
Comment (limebell) [inline NineChronicles.Headless/Controllers/GraphQLController.cs:287]: Is it okay to return only `Unauthorized` and `null`?

Comment (moreal) [reply NineChronicles.Headless/Controllers/GraphQLController.cs:287]: At least, it seems okay in this method to me. Can there be a problem?

Comment (limebell) [reply NineChronicles.Headless/Controllers/GraphQLController.cs:287]: `null` means "authorized"?

Comment (moreal) [reply NineChronicles.Headless/Controllers/GraphQLController.cs]: No, it doesn't. I refactored it like 7615507. How about these changes?

File: NineChronicles.Headless/Controllers/GraphQLController.cs [hunk 4]
```diff
@@ -106,14 +122,19 @@ public IActionResult SetPrivateKey([FromBody] SetPrivateKeyRequest request)
...
+
             if (StandaloneContext.NineChroniclesNodeService is null)
             {
                 // Waiting node service.
                 return new StatusCodeResult(StatusCodes.Status409Conflict);
             }
 
             bool mine = request.Mine;
-            if (mine)
+            if (mine)
```
Comment (riemannulus) [inline NineChronicles.Headless/Controllers/GraphQLController.cs:137]: Is this intented?

**Review by riemannulus — APPROVED** [review]

File: NineChronicles.Headless/Controllers/GraphQLController.cs (hunk 4 above)
Comment (moreal) [reply NineChronicles.Headless/Controllers/GraphQLController.cs]: No, it isn't.  I amended it, thanks 🙏

File: NineChronicles.Headless/Controllers/GraphQLController.cs (hunk 3 above)
Comment (limebell) [reply NineChronicles.Headless/Controllers/GraphQLController.cs]: It make more sense to me :)

**Pull request merged**